packages/portfolio-contract/src/portfolio.exo.ts

```diff
@@ -39,195 +45,187 @@ export const DECODE_CONTRACT_CALL_RESULT_ABI = [
-  receiveUpcall: M.call(M.record()).returns(M.promise()),
-});
+type AxelarGmpManager = {
+  localAccount: LocalAccount;
+  remoteAddressVK: VowKit<`0x${string}`>;
+};
 
 type EVMProtocolState = {
   chain: CaipChainId;
-  remoteAddressVK: VowKit<`0x${string}`>;
 };
 
-const EthL1: CaipChainId = 'eip155:1';
-
-/** ICA on Noble */
-type NobleAccount = OrchestrationAccount<{ chainId: 'noble-any' }>;
-
 type PortfolioKitState = {
-  localAccount: LocalAccount;
   Aave: EVMProtocolState | undefined;
   Compound: EVMProtocolState | undefined;
   USDN: NobleAccount | undefined;
+  Gmp: AxelarGmpManager | undefined;
 };
 
-// NOTE: This is host side code; can't use await.
 export const preparePortfolioKit = (
   zone: Zone,
   {
-    zcf,
-    vowTools,
+    axelarChainsMap,
     rebalance,
+    timer,
     proposalShapes,
+    vowTools,
+    zcf,
   }: {
-    zcf: ZCF;
-    vowTools: VowTools;
+    axelarChainsMap: AxelarChainsMap;
     rebalance: (
       seat: ZCFSeat,
       offerArgs: OfferArgsFor['rebalance'],
       keeper: unknown, // XXX avoid circular reference
     ) => Vow<any>; // XXX HostForGuest???
+    timer: Remote<TimerService>;
     proposalShapes: ReturnType<typeof makeProposalShapes>;
+    vowTools: VowTools;
+    zcf: ZCF;
   },
 ) =>
   zone.exoClassKit(
     'Portfolio',
     {
+      tap: M.interface('tap', {
+        receiveUpcall: M.call(M.record()).returns(M.promise()),
+      }),
       keeper: KeeperI,
       rebalanceHandler: OfferHandlerI,
       invitationMakers: M.interface('invitationMakers', {
         Rebalance: M.callWhen().returns(InvitationShape),
       }),
-      tap: EvmTapI,
     },
-
-    (localAccount: LocalAccount): PortfolioKitState => ({
-      localAccount,
-      Aave: undefined,
-      Compound: undefined,
-      USDN: undefined,
-    }),
+    (): PortfolioKitState => {
+      return {
+        Aave: undefined,
+        Compound: undefined,
+        USDN: undefined,
+        Gmp: undefined,
+      };
+    },
     {
       tap: {
         async receiveUpcall(event: VTransferIBCEvent) {
           trace('receiveUpcall', event);
 
-          const tx: FungibleTokenPacketData = JSON.parse(
-            atob(event.packet.data),
-          );
+          // TODO: dont use the same LCA for sending transfers to noble
+          let tx: FungibleTokenPacketData;
+          try {
+            tx = JSON.parse(atob(event.packet.data));
+          } catch (err) {
+            trace(
+              'Failed to parse packet data JSON in receiveUpcall:',
+              err,
+              event.packet.data,
+            );
+            return;
+          }
 
           trace('receiveUpcall packet data', tx);
           if (!tx.memo) return;
-          const memo: AxelarGmpIncomingMemo = JSON.parse(tx.memo); // XXX unsound! use typed pattern
-
-          if (!(memo.source_chain in PositionChain)) {
-            console.warn('unknown source_chain', memo);
+          let memo: AxelarGmpIncomingMemo;
```

Comment:
Do you know what I mean by `unsound!` below? We don't have conclusive evidence that the data coming out of JSON.parse is a `AxelarGmpIncomingMemo`. I **suggest**:

```suggestion
          let memo: unknown;
```